### ISSUE TITLE
Cygwin support

### ISSUE THREAD

**Issue by k5jat** [stsp/dosemu2#292]
I would like to see cygwin support. Since 64-bit windows no longer supports 16-bit dos programs, and official dosbox is limited to mostly game related features and not under active development, this would be a boon for those of us still needing dos support on modern Windows systems.

**Comment by stsp** [comment]
It would indeed be a very interesting
experiment, but dosemu by default uses
KVM for vm86, which is unlikely to be available
under cygwin. So it will have to resort to jit
emulation which will be only somewhat faster
than dosbox for realmode programs.
As for protected-mode programs, dosemu
uses many linux-specific features which may
be quite difficult to port to cygwin.
In either case, have you tried building it?
It would be interesting to see what build errors
are thrown.

**Comment by stsp** [comment]
You can help this by providing the qemu
image with the development environment
set up. I don't have windows installations
around to try myself. dosemu2 have a very
small dev power, so if you want the project
as big as this one to happen, you'll have to
offer some help. :)

**Comment by jschwartzenberg** [comment]
I can imagine it would be easier to use the new WSL subsystem instead of Cygwin to support Windows users. In theory Microsoft could even expose a working /dev/kvm device inside of that.

**Comment by runderwo** [comment]
There's also coLinux, which works without modification, but using vm86 on it resulted in a bluescreen last time I tried (~2011).  https://retro926.wordpress.com/2016/10/27/colinux-on-windows-10-part-1-introduction/

**Comment by bolle732** [comment]
I've used QEMU on Windows about 10 years ago. There even have been 2 kernel modules for Windows, to accelerate the speed. One from Fabrice itself and later another one from the open source community. But I can't tell, if it would be possible to run KVM itself as a hypervisor on Windows.

Sure, would be great to see it if someone(s) has/have the time to get that up. Could increase the user base of DOSEMU, for the good and/or bad...

**Comment by stsp** [comment]
@runderwo is colinux still active?
I really think cygwin/wsl would be a better
choice as it won't depend on some windows
ring0 drivers, which if what colinux AFAIK is.

So if people are so interested in this, someone
should start preparing the dev environment and
seeng what fails. I am sure we can get the basic
realmode support rather quickly.

**Comment by jschwartzenberg** [comment]
No coLinux is not active anymore and only works on 32-bit Windows versions which also still have an NTVDM which works without CPU emulation.

If you really want an integrated experience without CPU emulation, I would suggest asking Microsoft for /dev/kvm support in WSL so that dosemu can be supported. If you don't care about integration, it's likely possible to run DOS within the hypervisor that Windows includes. If you don't care about emulation, there is a special version of DOSBox with similar integration features and there's the unofficial NTVDM port. I don't use Windows myself though, so I didn't try any of these.

**Comment by stsp** [comment]
> /dev/kvm support in WSL so that dosemu can be supported.

We don't really need this for the start.
Realmode works fine with cpu-emu and
for DPMI we don't use kvm (yet) either,
so its not a big deal. I think DPMI will require
some porting efforts but realmode should
probably work.

> is a special version of DOSBox with similar integration features

What integration does dosbox need?

**Comment by jschwartzenberg** [comment]
> What integration does dosbox need?

These people found it lacking:
https://www.vdos.info/
http://www.vdosplus.org/

**Comment by jschwartzenberg** [comment]
Currently the build in Cygwin fails with:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -I. -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o dosctrl.o dosctrl.c
dosctrl.c:30:20: fatal error: sys/vt.h: No such file or directory
compilation terminated.
make[2]: *** [../../../src/Makefile.common:33: dosctrl.o] Error 1
```

If relying on just emulation, shouldn't it be possible to compile on ARM for instance as well?

**Comment by andrewbird** [comment]
> If relying on just emulation, shouldn't it be possible to compile on ARM for instance as well?

Mmm, depends on what ARM platforms, I guess there may be issues with disk structs etc if targetting a big endian machine.

**Comment by stsp** [comment]
> Currently the build in Cygwin fails with:

You can simply remove dosctrl from src/tools/periph/Makefile

> big endian machine.

Yes, certainly.
Plus jit will fail, and DPMI will be completely unportable
(rather than challenging, as I suspect it is with cygwin).
Which leaves us with just a vm86sim?
So targeting arm looks much less feasible. And what't
the use if dosbox does very good job there already?

**Comment by stsp** [comment]
> Currently the build in Cygwin fails with:

This should now be fixed.

@andrewbird could you please remind me
why do we still need mkhdimage?

**Comment by andrewbird** [comment]
Only to allow a user to build a virgin, unformatted disk image. I use it myself when testing out versions of DOS to check how the native tools format disks.

**Comment by jschwartzenberg** [comment]
Next failure:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o text.o text.c
In file included from text.c:37:0:
../../../src/include/emu.h:14:21: fatal error: syscall.h: No such file or directory
compilation terminated.
make[2]: *** [../../../src/Makefile.common:33: text.o] Error 1
```

**Comment by stsp** [comment]
OK, someone needs to start submitting patches
for this. :) This particular one is simple (put ifdefs
around gettid), but I suspect there are many more to follow.
Google shows that while in the past there was a way
to cross-compile with cygwin on linux, now this probably
doesn't work (or maybe it does with wine).
If someone manages to set up the cross-compilation
toolchain, please report your findings here. ;)

**Comment by stsp** [comment]
> Next failure:

Should now be fixed.

**Comment by jschwartzenberg** [comment]
Next failure:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o text.o text.c
In file included from ../../../src/include/emu.h:15:0,
                 from text.c:37:
../../../src/include/machcompat.h:32:18: fatal error: mach.h: No such file or directory
compilation terminated.
make[2]: *** [../../../src/Makefile.common:33: text.o] Error 1
```

**Comment by stsp** [comment]
Should now be fixed.

**Comment by jschwartzenberg** [comment]
Next failure:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o text.o text.c
In file included from ../../../src/include/emu.h:15:0,
                 from text.c:37:
../../../src/include/cpu.h:433:42: warning: ‘struct sigcontext’ declared inside parameter list
 int cpu_trap_0f (unsigned char *, struct sigcontext *);
                                          ^
../../../src/include/cpu.h:433:42: warning: its scope is only this definition or declaration, which is probably not what you want
In file included from ../../../src/include/emu.h:17:0,
                 from text.c:37:
../../../src/include/mouse.h:207:33: warning: ‘struct vm86_regs’ declared inside parameter list
 extern void mouse_helper(struct vm86_regs *);
                                 ^
In file included from text.c:37:0:
../../../src/include/emu.h:32:23: warning: ‘struct sigcontext’ declared inside parameter list
 int vm86_fault(struct sigcontext *scp);
                       ^
../../../src/include/emu.h:65:22: error: field ‘vm86ps’ has incomplete type
   struct vm86_struct vm86ps;
                      ^
../../../src/include/emu.h:66:26: error: invalid application of ‘sizeof’ to incomplete type ‘struct vm86_struct’
   unsigned char b[sizeof(struct vm86_struct)];
                          ^
../../../src/include/emu.h:67:27: error: invalid application of ‘sizeof’ to incomplete type ‘struct vm86_struct’
   unsigned short w[sizeof(struct vm86_struct)/2];
                           ^
../../../src/include/emu.h:68:25: error: invalid application of ‘sizeof’ to incomplete type ‘struct vm86_struct’
   unsigned int d[sizeof(struct vm86_struct)/4];
                         ^
../../../src/include/emu.h:69:26: error: field ‘vm86compat’ has incomplete type
   struct vm86plus_struct vm86compat;
                          ^
In file included from text.c:37:0:
../../../src/include/emu.h:394:30: warning: ‘struct vm86_regs’ declared inside parameter list
 extern int mfs_helper(struct vm86_regs *regs);
                              ^
../../../src/include/emu.h:396:30: warning: ‘struct vm86_regs’ declared inside parameter list
 extern short pop_word(struct vm86_regs *);
                              ^
../../../src/include/emu.h:458:26: warning: ‘struct vm86_regs’ declared inside parameter list
 extern int ems_fn(struct vm86_regs *);
                          ^
make[2]: *** [../../../src/Makefile.common:33: text.o] Error 1
```

**Comment by stsp** [comment]
Partially fixed.
Seems we need to provide "struct sigcontext"
ourselves in this case. Perhaps copy/pasting the
stripped header from asm/sigcontext.h would do.
Anyone? :)

**Comment by andrewbird** [comment]
Copying struct sigcontext would be fine, but it should only be used for the non linux case IMHO. Using system uapi headers for types that are used in interaction with the kernel is always preferable.

**Comment by stsp** [comment]
Of course you can check for CYGWIN
or WIN32 or alike. This is not a problem.

Another possibility may be to look into
libsigsegv. But this is definitely a bigger
task than just rip off the header.

**Comment by jschwartzenberg** [comment]
I installed libsigsegv, now I get this error:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o text.o text.c
In file included from ../../../src/include/cpu.h:16:0,
                 from ../../../src/include/emu.h:15,
                 from text.c:37:
../../../src/include/vm86_compat.h:7:33: fatal error: asm/processor-flags.h: No such file or directory
compilation terminated.
make[2]: *** [../../../src/Makefile.common:33: text.o] Error 
```

**Comment by stsp** [comment]
> I installed libsigsegv

This isn't needed, we don't currently use it.
I thought maybe it will provide some sigcontext
wrapper but it doesn't.

> now I get this error:

Should now be fixed.

**Comment by jschwartzenberg** [comment]
Next failure:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -fplan9-extensions -Wno-maybe-uninitialized   -g -O2 -o instremu.o instremu.c
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
../../../src/include/cpu.h:429:42: warning: ‘struct sigcontext’ declared inside parameter list
 int cpu_trap_0f (unsigned char *, struct sigcontext *);
                                          ^
../../../src/include/cpu.h:429:42: warning: its scope is only this definition or declaration, which is probably not what you want
In file included from instremu.c:49:0:
../../../src/include/emu.h:32:23: warning: ‘struct sigcontext’ declared inside parameter list
 int vm86_fault(struct sigcontext *scp);
                       ^
In file included from ../../../src/include/../dosext/dpmi/dpmi.h:5:0,
                 from ../../../src/include/dpmi.h:6,
                 from instremu.c:51:
../../../src/include/emu-ldt.h:10:21: fatal error: asm/ldt.h: No such file or directory
compilation terminated.
make[2]: *** [../../../src/Makefile.common:33: instremu.o] Error 1
```

**Comment by stsp** [comment]
I already have the sigcontext header stripped for dosemu:
https://github.com/stsp/pmdapi/blob/master/src/sigcontext.h
Can someone just put it in, include under proper ifdef
and test that it goes further?

**Comment by jschwartzenberg** [comment]
I included that header in instremu.c, but then I get this:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o instremu.o instremu.c
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
../../../src/include/cpu.h:429:42: warning: ‘struct sigcontext’ declared inside parameter list
 int cpu_trap_0f (unsigned char *, struct sigcontext *);
                                          ^
../../../src/include/cpu.h:429:42: warning: its scope is only this definition or declaration, which is probably not what you want
In file included from instremu.c:49:0:
../../../src/include/emu.h:32:23: warning: ‘struct sigcontext’ declared inside parameter list
 int vm86_fault(struct sigcontext *scp);
                       ^
In file included from ../../../src/include/dpmi.h:6:0,
                 from instremu.c:51:
../../../src/include/../dosext/dpmi/dpmi.h:104:21: error: field ‘stack_frame’ has incomplete type
   struct sigcontext stack_frame;
                     ^
In file included from ../../../src/include/../dosext/dpmi/dpmi.h:5:0,
                 from ../../../src/include/dpmi.h:6,
                 from instremu.c:51:
../../../src/include/emu-ldt.h:22:23: error: ‘LDT_ENTRIES’ undeclared here (not in a function)
 #define MAX_SELECTORS LDT_ENTRIES
                       ^
../../../src/include/../dosext/dpmi/dpmi.h:143:25: note: in expansion of macro  MAX_SELECTORS’
 extern SEGDESC Segments[MAX_SELECTORS];
                         ^
In file included from ../../../src/include/dpmi.h:6:0,
                 from instremu.c:51:
../../../src/include/../dosext/dpmi/dpmi.h:148:47: error: ‘LDT_ENTRY_SIZE’ undeclared here (not in a function)
 extern unsigned char ldt_buffer[LDT_ENTRIES * LDT_ENTRY_SIZE];
                                               ^
In file included from instremu.c:55:0:
../../../src/include/sigcontext.h:33:8: error: redefinition of ‘struct _fpstate’
 struct _fpstate {
        ^
In file included from /usr/include/sys/signal.h:112:0,
                 from /usr/include/signal.h:6,
                 from ../../../src/include/cpu.h:19,
                 from ../../../src/include/emu.h:15,
                 from instremu.c:49:
/usr/include/cygwin/signal.h:34:8: note: originally defined here
 struct _fpstate
        ^
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
instremu.c: In function ‘scp_to_x86_regs’:
../../../src/include/cpu.h:388:21: error: ‘struct sigcontext’ has no member named ‘rax’
 #define _rax    (scp->rax)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:418:24: note: in expansion of macro ‘_rax’
 #define _eax    DWORD_(_rax)
                        ^
instremu.c:2343:16: note: in expansion of macro ‘_eax’
     x86->eax = _eax;
                ^
../../../src/include/cpu.h:385:21: error: ‘struct sigcontext’ has no member named ‘rbx’
 #define _rbx    (scp->rbx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:413:24: note: in expansion of macro ‘_rbx’
 #define _ebx    DWORD_(_rbx)
                        ^
instremu.c:2344:16: note: in expansion of macro ‘_ebx’
     x86->ebx = _ebx;
                ^
../../../src/include/cpu.h:387:21: error: ‘struct sigcontext’ has no member named ‘rcx’
 #define _rcx    (scp->rcx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:415:24: note: in expansion of macro ‘_rcx’
 #define _ecx    DWORD_(_rcx)
                        ^
instremu.c:2345:16: note: in expansion of macro ‘_ecx’
     x86->ecx = _ecx;
                ^
../../../src/include/cpu.h:386:21: error: ‘struct sigcontext’ has no member named ‘rdx’
 #define _rdx    (scp->rdx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:414:24: note: in expansion of macro ‘_rdx’
 #define _edx    DWORD_(_rdx)
                        ^
instremu.c:2346:16: note: in expansion of macro ‘_edx’
     x86->edx = _edx;
                ^
../../../src/include/cpu.h:382:21: error: ‘struct sigcontext’ has no member named ‘rsi’
 #define _rsi    (scp->rsi)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:410:24: note: in expansion of macro ‘_rsi’
 #define _esi    DWORD_(_rsi)
                        ^
instremu.c:2347:16: note: in expansion of macro ‘_esi’
     x86->esi = _esi;
                ^
../../../src/include/cpu.h:381:21: error: ‘struct sigcontext’ has no member named ‘rdi’
 #define _rdi    (scp->rdi)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:409:24: note: in expansion of macro ‘_rdi’
 #define _edi    DWORD_(_rdi)
                        ^
instremu.c:2348:16: note: in expansion of macro ‘_edi’
     x86->edi = _edi;
                ^
../../../src/include/cpu.h:383:21: error: ‘struct sigcontext’ has no member named ‘rbp’
 #define _rbp    (scp->rbp)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:411:24: note: in expansion of macro ‘_rbp’
 #define _ebp    DWORD_(_rbp)
                        ^
instremu.c:2349:16: note: in expansion of macro ‘_ebp’
     x86->ebp = _ebp;
                ^
../../../src/include/cpu.h:384:21: error: ‘struct sigcontext’ has no member named ‘rsp’
 #define _rsp    (scp->rsp)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:412:24: note: in expansion of macro ‘_rsp’
 #define _esp    DWORD_(_rsp)
                        ^
instremu.c:2350:16: note: in expansion of macro ‘_esp’
     x86->esp = _esp;
                ^
../../../src/include/cpu.h:389:21: error: ‘struct sigcontext’ has no member named ‘rip’
 #define _rip    (scp->rip)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:419:24: note: in expansion of macro ‘_rip’
 #define _eip    DWORD_(_rip)
                        ^
instremu.c:2351:16: note: in expansion of macro ‘_eip’
     x86->eip = _eip;
                ^
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
../../../src/include/cpu.h:393:21: error: ‘struct sigcontext’ has no member named ‘__pad0’
 #define _ss     (scp->__pad0)
                     ^
instremu.c:2356:15: note: in expansion of macro ‘_ss’
     x86->ss = _ss;
               ^
../../../src/include/cpu.h:393:21: error: ‘struct sigcontext’ has no member named ‘__pad0’
 #define _ss     (scp->__pad0)
                     ^
instremu.c:2362:35: note: in expansion of macro ‘_ss’
     x86->ss_base = GetSegmentBase(_ss);
                                   ^
instremu.c:2365:26: warning: implicit declaration of function ‘dpmi_mhp_get_selector_size’ [-Wimplicit-function-declaration]
     x86->_32bit = _cs && dpmi_mhp_get_selector_size(_cs) ? 1 : 0;
                          ^
instremu.c:2365:5: warning: nested extern declaration of ‘dpmi_mhp_get_selector_size’ [-Wnested-externs]
     x86->_32bit = _cs && dpmi_mhp_get_selector_size(_cs) ? 1 : 0;
     ^
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
instremu.c: In function ‘x86_regs_to_scp’:
../../../src/include/cpu.h:393:21: error: ‘struct sigcontext’ has no member named ‘__pad0’
 #define _ss     (scp->__pad0)
                     ^
instremu.c:2403:5: note: in expansion of macro ‘_ss’
     _ss = x86->ss;
     ^
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
../../../src/include/cpu.h:388:21: error: ‘struct sigcontext’ has no member named ‘rax’
 #define _rax    (scp->rax)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:418:24: note: in expansion of macro ‘_rax’
 #define _eax    DWORD_(_rax)
                        ^
instremu.c:2404:5: note: in expansion of macro ‘_eax’
     _eax = x86->eax;
     ^
../../../src/include/cpu.h:385:21: error: ‘struct sigcontext’ has no member named ‘rbx’
 #define _rbx    (scp->rbx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:413:24: note: in expansion of macro ‘_rbx’
 #define _ebx    DWORD_(_rbx)
                        ^
instremu.c:2405:5: note: in expansion of macro ‘_ebx’
     _ebx = x86->ebx;
     ^
../../../src/include/cpu.h:387:21: error: ‘struct sigcontext’ has no member named ‘rcx’
 #define _rcx    (scp->rcx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:415:24: note: in expansion of macro ‘_rcx’
 #define _ecx    DWORD_(_rcx)
                        ^
instremu.c:2406:5: note: in expansion of macro ‘_ecx’
     _ecx = x86->ecx;
     ^
../../../src/include/cpu.h:386:21: error: ‘struct sigcontext’ has no member named ‘rdx’
 #define _rdx    (scp->rdx)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:414:24: note: in expansion of macro ‘_rdx’
 #define _edx    DWORD_(_rdx)
                        ^
instremu.c:2407:5: note: in expansion of macro ‘_edx’
     _edx = x86->edx;
     ^
../../../src/include/cpu.h:382:21: error: ‘struct sigcontext’ has no member named ‘rsi’
 #define _rsi    (scp->rsi)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:410:24: note: in expansion of macro ‘_rsi’
 #define _esi    DWORD_(_rsi)
                        ^
instremu.c:2408:5: note: in expansion of macro ‘_esi’
     _esi = x86->esi;
     ^
../../../src/include/cpu.h:381:21: error: ‘struct sigcontext’ has no member named ‘rdi’
 #define _rdi    (scp->rdi)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:409:24: note: in expansion of macro ‘_rdi’
 #define _edi    DWORD_(_rdi)
                        ^
instremu.c:2409:5: note: in expansion of macro ‘_edi’
     _edi = x86->edi;
     ^
../../../src/include/cpu.h:383:21: error: ‘struct sigcontext’ has no member named ‘rbp’
 #define _rbp    (scp->rbp)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:411:24: note: in expansion of macro ‘_rbp’
 #define _ebp    DWORD_(_rbp)
                        ^
instremu.c:2410:5: note: in expansion of macro ‘_ebp’
     _ebp = x86->ebp;
     ^
../../../src/include/cpu.h:384:21: error: ‘struct sigcontext’ has no member named ‘rsp’
 #define _rsp    (scp->rsp)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:412:24: note: in expansion of macro ‘_rsp’
 #define _esp    DWORD_(_rsp)
                        ^
instremu.c:2411:5: note: in expansion of macro ‘_esp’
     _esp = x86->esp;
     ^
../../../src/include/cpu.h:389:21: error: ‘struct sigcontext’ has no member named ‘rip’
 #define _rip    (scp->rip)
                     ^
../../../src/include/cpu.h:61:40: note: in definition of macro ‘DWORD_’
 #define DWORD_(wrd) (((union dword *)&(wrd))->d)
                                        ^
../../../src/include/cpu.h:419:24: note: in expansion of macro ‘_rip’
 #define _eip    DWORD_(_rip)
                        ^
instremu.c:2412:5: note: in expansion of macro ‘_eip’
     _eip = x86->eip;
     ^
m
```

**Comment by stsp** [comment]
Try including it into sig.h instead.
If doesn't help, then in emu.h.

**Comment by jschwartzenberg** [comment]
Included in sig.h:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o instremu.o instremu.c
In file included from ../../../src/include/emu.h:15:0,
                 from instremu.c:49:
../../../src/include/cpu.h:429:42: warning: ‘struct sigcontext’ declared inside parameter list
 int cpu_trap_0f (unsigned char *, struct sigcontext *);
                                          ^
../../../src/include/cpu.h:429:42: warning: its scope is only this definition or declaration, which is probably not what you want
In file included from instremu.c:49:0:
../../../src/include/emu.h:33:23: warning: ‘struct sigcontext’ declared inside parameter list
 int vm86_fault(struct sigcontext *scp);
                       ^
In file included from ../../../src/include/../dosext/dpmi/dpmi.h:5:0,
                 from ../../../src/include/dpmi.h:6,
                 from instremu.c:51:
../../../src/include/emu-ldt.h:10:21: fatal error: asm/ldt.h: No such file or directory
compilation terminated.
```
Included in emu.h:
```
gcc -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2" -I../../../src/include -I../../../src/plugin/include  -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -o text.o text.c
In file included from ../../../src/include/emu.h:15:0,
                 from text.c:37:
../../../src/include/cpu.h:429:42: warning: ‘struct sigcontext’ declared inside parameter list
 int cpu_trap_0f (unsigned char *, struct sigcontext *);
                                          ^
../../../src/include/cpu.h:429:42: warning: its scope is only this definition or declaration, which is probably not what you want
In file included from ../../../src/include/emu.h:19:0,
                 from text.c:37:
../../../src/include/sigcontext.h:33:8: error: redefinition of ‘struct _fpstate’
 struct _fpstate {
        ^
In file included from /usr/include/sys/signal.h:112:0,
                 from /usr/include/signal.h:6,
                 from ../../../src/include/cpu.h:19,
                 from ../../../src/include/emu.h:15,
                 from text.c:37:
/usr/include/cygwin/signal.h:34:8: note: originally defined here
 struct _fpstate
```

**Comment by jschwartzenberg** [comment]
About supporting other platforms, it appears there's a hypervisor in Mac OS X nowadays and someone wrote a DOS emulator for it: https://github.com/mist64/hvdos
I don't have any Mac stuff or experience, but I suppose that in theory dosemu could be ported to Mac OS X as well.

**Comment by stsp** [comment]
Just 4 commits.
I don't think it works at all (well, it is
said it can run a few .com progs).
All it does is to set an unrestricted
guest mode with vtx and runs a simple
.com loader there with a couple of
hand-written DOS interrupt handlers.
So from this lesson we know Mac OS X
has a low-level vtx API for which we
can write a backend. But its not something
that can be done in an hour.
Personally I think we don't have dev
resources to spend on other platforms -
we strive to get something functional at
least on linux. :)

**Comment by stsp** [comment]
> If relying on just emulation, shouldn't it be possible to compile on ARM for instance as well?

Just out of curiosity, I tried that on my
arm phone. After fiddling with realpath
a bit (which is now removed), I managed
to get configure completed. But the build
fails very quickly, namely, when it tries to
build the .com and .sys files from asm.
Supporting this properly will require the
cross binutils together with the native
binutils for C code. I refuse to code this
up in autoconf+makefiles. :)

**Comment by JL2210** [comment]
Could you cross-compile them and copy it over?

**Comment by stsp** [comment]
> I know ld has an option -melf_i386

Well, who told you it should work on ARM?
```
$ ld -melf_i386
ld: unrecognised emulation mode: elf_i386
Supported emulations: armelf_linux_eabi armelfb_linux_eabi
```

Please, senseless comments are not going to help.

**Comment by volkertb** [comment]
Man, it seems like it would be quite useful (to DOSEMU2, but potentially to other projects as well) if there were some kind of API that would provide a cross-platform abstraction on top of KVM in Linux, WHPX in Windows, Hypervisor.framework on macOS, and optionally HAXM as well. Basically like QEMU, but just the lower-level hypervisor-control part. Not unlike how MoltenVK provides a Vulkan-compatible layer on top of the Metal API in macOS.

Does that sound crazy?

**Comment by jschwartzenberg** [comment]
> Man, it seems like it would be quite useful (...) if there were some kind of API that would provide a cross-platform abstraction on top of KVM in Linux,

I think that would be useful for VirtualBox. It still lacks KVM support.

**Comment by stsp** [comment]
KVM is already an API for low-level
hypervisor control. I am not sure what
kind of API you want to define. We
could define vm86() on top of KVM,
for example, and this is what we already
do.

**Comment by stsp** [comment]
As for cygwin, can someone please
re-try and see where it breaks now?
We have a much tighter configure
right now, so I guess if cygwin misses
something that we need, there will
be no compilation errors as before,
but rather a configure error.
But the code is now much more portable.

**Comment by Nobody-Special** [comment]
I'll give it a shot.
No promises on the date though.

**Comment by stsp** [comment]
Thanks.
The freebsd port is nearly completed,
so within 2-3 days it would be the right
time to try.

**Comment by Nobody-Special** [comment]
I won't have access to a Windows machine over the weekend, so I guess I'll try on monday.

**Comment by stsp** [comment]
That gives me more time to complete
the bsd port. :) I am quite certain after
that, the cygwin port won't be a problem.

**Comment by stsp** [comment]
I installed cygwin under linux and found
this bug:
https://sourceware.org/bugzilla/show_bug.cgi?id=30561
So unfortunately cygwin support shoud
be postponed, lets monitor that bug and
hope it will soon be fixed.

**Comment by stsp** [comment]
Also if someone is going to try building
under windows, there is an additional
complication which is the need to install
cross-binutils. Presumably from here:
https://github.com/lordmilko/i686-elf-tools/releases

**Comment by stsp** [comment]
Yet another complication is building fdpp
under windows. That would require clang,
presumably from here:
https://github.com/mstorsjo/llvm-mingw/releases
I would suggest trying that first, as long as
building of dosemu2 itself is blocked by the
cygwin bug. Building fdpp would already be
a big step.

**Comment by Nobody-Special** [comment]
I've tried compiling fdpp with clang(++) available on Cygwin, but this crazy old version of Clang has no PIC command line flag for windows-unknown-cygwin.

Tomorrow I'll try building with llvm-mingw which you have linked, but it'll be my first time tinkering with tool-chain on Windows (I've been on Linux since I was 10 years old), so the results might arrive later :)

Ps. What is more appropriate. Creating new tracking issue on dosemu2/fdpp or appending this issue?

**Comment by stsp** [comment]
All projects I linked, have the binary releases.
But I am not sure about -fpic actually.
Basically under windows its probably not
possible to create libfdpp.so.xx.yy.
Probably instead we'd need to create
libfdpp.dll, which may require some other
set of options than -fpic.
Or does the cygwin allow creating some
sort of a solib?

**Comment by stsp** [comment]
Yes, fdpp-specific things should be opened
at fdpp tracker.

**Comment by stsp** [comment]
So please hack away -fpic and rename
.so to .dll in a makefile. Maybe it will then
be better.

**Comment by stsp** [comment]
You can now compile dosemu2 for
cygwin using configure-cyg64 script.
comcom64 is probably not possible
to port, so better use pre-built command.com.
fdpp may be feasible, not sure.
